### PR TITLE
feat: add social media sharing

### LINKED ISSUE #61: Share app posts to social media sites
Design Ref: 
- https://xd.adobe.com/view/dd5d0245-b92b-4678-9d4a-48b3a6f48191-880e/screen/f5f3abf2-9384-460a-bac7-544c49648c0c/specs/
- https://xd.adobe.com/view/dd5d0245-b92b-4678-9d4a-48b3a6f48191-880e/screen/09ac0874-4933-4982-9e2e-b5d2c82d6c5f/specs/

A user should be allowed to share the content from the feed to social media sites.

**Initial Approach**
- Add `share_plus` (https://pub.dev/packages/share_plus)
- Ensure the content makes it over to the plugin for posting on the socials

**Edit**
Looks like we may need `appinio_social_share` (https://pub.dev/packages/appinio_social_share) to really achieve the social media effect

### LINKED ISSUE #88: Weird loading on `Share`
As can be observed, when I click the Share button. All the posts are loading. This is quite confusing.

### PR DESCRIPTION
**Pull request by DanroyMwangi**
## Description

This PR:
  - Implements a social media bottom sheet accessible via the share button on the feed page.
  - Adds a reusable social media button widget for use within the bottom sheet.
  - Enables users to share content on various platforms, including Telegram, Twitter, WhatsApp, and Facebook.

Closes #61 
Closes #88 

### Areas for Future Improvement

**Error Handling:** 
Currently, errors for missing social media apps are displayed in the terminal. Consider enhancing the user experience by showing these errors directly in the UI.
Screenshot from 2024-08-16 23-11-08

## Type of Change

- [x] ✨ New feature (non-breaking change that adds functionality)
- [x] 🛠️ Bug fix (non-breaking change that fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

## Test on Device

You can download the APK from Firebase App Distribution once this PR builds successfully via the following link: [Download APK](https://appdistribution.firebase.dev/i/c796669942f8a811)

## Look
  - https://github.com/user-attachments/assets/1969fbd6-a902-4872-b997-2c8059557ce9
  - photo_2024-08-16_23-32-39

### CODE REVIEW

File: pubspec.yaml
Comment (MillerAdulu) [inline pubspec.yaml:1]: Please bump the app version: `version: 1.9.01+1901`.

File: lib/features/feed/cubit/share_feed_post_cubit.dart [hunk 1]
```diff
@@ -1,30 +1,104 @@
...
-      emit(ShareFeedPostState.error(e.toString()));
+      emit(ShareFeedPostState.error('Failed to share post: $e'));
+    }
+  }
+
+  Future<void> _shareToTelegram(String message, String? filePath) async {
+    try {
+      await appinioSocialShare.android.shareToTelegram(message, filePath);
+    } catch (e) {
+      throw Exception('Error sharing to Telegram: $e');
```
Comment (MillerAdulu) [inline lib/features/feed/cubit/share_feed_post_cubit.dart:77]: Throw `Failure` instead. 

Sample: `lib/common/utils/network.dart:75`

Then in the calling function, catch it like so `lib/features/about/cubit/fetch_individual_organisers_cubit.dart:32-34`

File: lib/features/feed/cubit/share_feed_post_cubit.dart [hunk 2]
```diff
@@ -1,30 +1,104 @@
...
+    }
+  }
+
+  Future<void> _shareToTwitter(String message, String? filePath) async {
+    try {
+      await appinioSocialShare.android.shareToTwitter(message, filePath);
+    } catch (e) {
+      throw Exception('Error sharing to Twitter: $e');
     }
   }
```
Comment (MillerAdulu) [inline lib/features/feed/cubit/share_feed_post_cubit.dart:103]: Please append the Fluttercon website to the message, like so `\n\nhttps://fluttercon.dev'` as it was prior

File: lib/features/feed/cubit/share_feed_post_cubit.dart (hunk 1 above)
Comment (MillerAdulu) [reply lib/features/feed/cubit/share_feed_post_cubit.dart:77]: On second thought, you don't have to actually throw anything here. Let me pull your branch and make suggestions after I have seen the code on my local.

**Review by MillerAdulu — CHANGES_REQUESTED** [review]
Please share if there are any iOS-specific steps that need to be applied so that I can have them noted down. This app currently runs smoothly on both platforms.

**Comment by MillerAdulu** [comment]
@DanroyMwangi,

Quite a number of formatting and lining issues, do ensure to run the `fmt` command in the `Makefile` once you are done so that the pipeline check is passed.

File: lib/features/feed/cubit/platform.dart
Comment (MillerAdulu) [inline lib/features/feed/cubit/platform.dart:1]: Move this file to `lib/common/data/enums`

File: lib/features/feed/cubit/share_feed_post_cubit.dart (hunk 2 above)
Comment (MillerAdulu) [reply lib/features/feed/cubit/share_feed_post_cubit.dart:103]: Scratch [this](https://github.com/droidconKE/flutterconKEApp/pull/91#discussion_r1720326983). Seen what you did.

Comment (MillerAdulu) [inline lib/features/feed/cubit/share_feed_post_cubit.dart:103]: The best place to have these individual functions would be a repository. Think of it this way. 

Repos - Direct interaction with the platform
Cubits - Updating the UI with the state
Widget - Results of the operations

The sharing function may need to be used in another instance, `lib/features/sessions/ui/session_details/session_details.dart` so best to have that independent of which screen it's used in. Please make a repo file for the functions like so:
- In `lib/common/repository`, make `share_repository.dart`
- Inside, define the class `class ShareRepository {` and ensure to annotate it with `@singleton` as seen in `lib/common/repository/api_repository.dart:10`. Run code gen `gen:` command in the `Makefile`. This ensures that the class is available for injection
- Then you can inject it into `ShareFeedPostCubit`. You can see a sample of the injection of other cubits in  the `providers` array `lib/bootstrap.dart:54`
- Then inside the cubit, you have a much neater file. 
- In the `api_repository.dart` file, you can observe how the try...catch happens. If you want to throw a particular error, throw the `Failure (lib/common/data/models/failure.dart)` then you can ensure to catch it in the cubit. Sample catch: `lib/features/about/cubit/fetch_individual_organisers_cubit.dart:32-34`

**Review by MillerAdulu — CHANGES_REQUESTED** [review]

**Comment by MillerAdulu** [comment]
Overall, great approach to solving the issue. We should be able to finalize after these minor changes.

**Comment by DanroyMwangi** [comment]
> Please share if there are any iOS-specific steps that need to be applied so that I can have them noted down. This app currently runs smoothly on both platforms.

I made changes for the Android platform based on the [Appinio Social Share documentation](https://pub.dev/packages/appinio_social_share). Unfortunately, I was unable to test these changes on iOS.

File: pubspec.yaml
Comment (DanroyMwangi) [reply pubspec.yaml:1]: Done.

File: lib/features/feed/cubit/platform.dart
Comment (DanroyMwangi) [reply lib/features/feed/cubit/platform.dart:1]: Let me do this right now.

File: lib/features/feed/cubit/share_feed_post_cubit.dart (hunk 2 above)
Comment (DanroyMwangi) [reply lib/features/feed/cubit/share_feed_post_cubit.dart:103]: I will work on this first thing in the morning.

**Comment by MillerAdulu** [comment]
> > Please share if there are any iOS-specific steps that need to be applied so that I can have them noted down. This app currently runs smoothly on both platforms.
> 
> I made changes for the Android platform based on the [Appinio Social Share documentation](https://pub.dev/packages/appinio_social_share). Unfortunately, I was unable to test these changes on iOS.

What particular changes were made for iOS to allow me to test them on my end? Just the exact lines that need changing since we don't support all the platforms listed there.

**Comment by MillerAdulu** [comment]
@DanroyMwangi 

Do you have a time estimate for this. Needing to assign another issue that's dependent on this being merged to avoid unpleasant merge conflicts.

**Comment by DanroyMwangi** [comment]
> @DanroyMwangi 
> 
> Do you have a time estimate for this. Needing to assign another issue that's dependent on this being merged to avoid unpleasant merge conflicts. 

Hey, I’m currently a bit tied up. The earliest I can deliver this is by this evening. I appreciate your patience!

**Comment by MillerAdulu** [comment]
Made a new issue #94 to enable Facebook sharing to happen on iOS and Android. Will finalize the pending Facebook work on that issue to allow work to move on. Thank you for your contribution merged via #93

**Pull request closed without merging**